lte/gateway/c/session_manager/SetMessageManagerHandler.cpp

```diff
@@ -115,8 +122,70 @@ void SetMessageManagerHandler::SetAmfSessionContext(
+      }
+    }
+    response_callback(Status::OK, SmContextVoid());
+  });
+}
+
+/* Handling set message from AMF
+ * check if PDU_SESSION exists
+ * if the PDU_SESSION doen't exists, log and ignore
+ * If EXISTING_PDU_SESSION, get the session entry, check on incoming session
+ * state and version  accordingly take action, write to memory by SessionStore
+ */
+
+void SetMessageManagerHandler::SetSmfNotification(
+    ServerContext* context, const SetSmNotificationContext* notif,
+    std::function<void(Status, SmContextVoid)> response_callback) {
+  auto& notif_cpy = *notif;
+  // Print the message from AMF
+  PrintGrpcMessage(static_cast<const google::protobuf::Message&>(notif_cpy));
+
+  // Requested message from AMF to release the session
+  m5g_enforcer_->get_event_base().runInEventBaseThread([this, response_callback,
+                                                        notif_cpy]() {
+    /* Read the proto message and check for state. Get the config out of proto.
+     * Code for Active state notification
+     */
+    // extract values from proto
+    auto pdu_id = notif_cpy.rat_specific_notification().pdu_session_id();
+    auto imsi       = notif_cpy.common_context().sid().id();
+    std::string dnn = notif_cpy.common_context().apn();
+    /* Read the SessionMap from global session_store */
+    SessionSearchCriteria criteria(imsi, IMSI_AND_PDUID, pdu_id);
+    auto session_map       = session_store_.read_sessions({imsi});
+    auto session_it        = session_store_.find_session(session_map, criteria);
+    std::string session_id = id_gen_.gen_session_id(imsi);
+    auto& session          = **session_it;
+    if (session_it) {
+      MLOG(MINFO) << "Received Session Notificaiton from UE with IMSI: " << imsi
+                  << " of session Id " << session->get_session_id();
+      if (notif_cpy.common_context().sm_session_version() == 0) {
+        MLOG(MERROR) << "Wrong version received from AMF for IMSI " << imsi;
+      } else {
+        if (notif_cpy.rat_specific_notification().request_type() ==
+            EXISTING_PDU_SESSION) {
+          auto session_update =
+              SessionStore::get_default_session_update(session_map);
+          auto session_id = session->get_session_id();
+          SessionStateUpdateCriteria& session_crit =
+              session_update[imsi][session_id];
+          m5g_enforcer_->m5g_state_change_action(
+              imsi, session_crit, notif_cpy, **session_it);
+          bool update_success = session_store_.update_sessions(session_update);
+          if (update_success) {
+            MLOG(MINFO) << "Successfully released and updated SessionStore "
+                        << "of subscriber" << imsi;
+          }
+        } else {
+          MLOG(MINFO) << " Wrong request type for sesion id"
+                      << session->get_session_id();
+        }
       }
+    } else {
+      MLOG(MINFO) << " No session found for IMSI: " << imsi << " DNN " << dnn;

```

Comment:
shouldn't you return an error if you don't find the session?

Reply:
I would also reverse all the ifs here. This way if you find any error you just exit. That would reduce all the levels of indentation you have there

Reply:
For the 1st comment, ? we shud return error where ?   Do u mean we shud change GRPC hanlder routine "SetSmfNotification" return type to void ? or "repsose_callback" as  failure instead of "Status::OK"

Reply:
Sorry I meant execute not return: 

this function always ends running `response_callback(Status::OK, SmContextVoid())`. No matter if the session was not found or if any other issue happened. Shouldn't this function run response callback with an error code in case of error? 

Also running an error and return would simplify the logic of the ifs making them one level only

Reply:
Thanks, let me do that

Reply:
Fixed, thanks.